### PR TITLE
Created model parsers and unit tests in net module

### PR DESCRIPTION
**Pull request by RachelTucker**

### CODE REVIEW

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java [hunk 1]
```diff
@@ -101,6 +101,7 @@ public static String toNetType(final String contractType) {
         }
     }
 
+    //TODO delete once removed from all templates
```
Comment (DenverM80) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java]: Is this TODO still valid?

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java [hunk 2]
```diff
@@ -192,6 +193,22 @@ public static String getEnumValues(
...
+    public static String commaSeparateStrings(
+            final ImmutableList<String> strings,
+            final int indent) {
+        if (isEmpty(strings)) {
+            return "";
+        }
+        return strings
+                .stream()
+                .map(i -> i)
+                .collect(Collectors.joining(",\n" + indent(indent)));
```
Comment (DenverM80) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java]: does this skip indenting the first element?

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java (hunk 1 above)
Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java]: Its still valid.  I wrote code to replace its functionality, but I did not do a complete swap because the pull request was big enough as is.  I am going to do this as a separate refactor.

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java (hunk 2 above)
Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java]: Yes. This way I can have the function call indented in the template. It makes the template little bit more readable imo

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java [hunk 3]
```diff
@@ -192,6 +193,22 @@ public static String getEnumValues(
...
+     */
+    public static String commaSeparateStrings(
+            final ImmutableList<String> strings,
+            final int indent) {
+        if (isEmpty(strings)) {
+            return "";
+        }
+        return strings
+                .stream()
+                .map(i -> i)
```
Comment (DenverM80) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java]: I don't think you need this map step if you're not transforming the input set

Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/NetHelper.java]: You are right. I will fix.

**Comment by DenverM80** [comment]
Overall this all looks good to me - the code is very readable, comments are in the right places to explain intent where its not immediately clear, and the test coverage is good as well.

File: ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java [hunk 4]
```diff
@@ -16,8 +16,8 @@
 package com.spectralogic.ds3autogen.api.models;
 
 public class Arguments {
-    final private String type;
```
Comment (rpmoore) [inline ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java]: Why was this changed from private to protected?

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import static com.spectralogic.ds3autogen.utils.Helper.uncapFirst;
+
+/**
+ * Represents an element within a .net model that contains
+ * a list of elements that have an encapsulating xml tag
+ */
+public class NullableEncapsulatedListElement  extends BaseNullableElement {
+
+    /** The encapsulating xml tag for the list described by this element */
+    protected final String encapsulatingTag;
```
Comment (rpmoore) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java]: Why is this protected and not made private with a public accessor?

File: ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java (hunk 4 above)
Comment (RachelTucker) [inline ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java]: Because NetNullableVariable extends Arguments, and I wanted access to the variables from the parent.

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java (hunk 5 above)
Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java]: No reason. It is never accessed from outside, so it doesn't need a getter.

File: ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java (hunk 4 above)
Comment (rpmoore) [inline ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java]: Why not add a public accessor?

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java (hunk 5 above)
Comment (rpmoore) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java]: It should be made private then.

File: ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java (hunk 4 above)
Comment (RachelTucker) [inline ds3-autogen-api/src/main/java/com/spectralogic/ds3autogen/api/models/Arguments.java]: will update

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java (hunk 5 above)
Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/NullableEncapsulatedListElement.java]: agreed

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/BaseNullableElement.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import static com.spectralogic.ds3autogen.utils.Helper.uncapFirst;
+
+/**
+ * Represents an element within a .net model parser whose data
+ * is within an xml tag
+ */
+public class BaseNullableElement implements NullableElement {
+
+    protected final String name;
```
Comment (rpmoore) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/BaseNullableElement.java]: It seems like these can be made private.

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/utils/NetNullableElementUtils.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+     */
+    public static NullableElement createNullableElement(
+            final String name,
+            final String type,
+            final String componentType,
+            final boolean nullable,
+            final String xmlTag,
+            final String encapsulatingXmlTag,
+            final boolean isAttribute) {
+        if (hasContent(componentType)) {
```
Comment (rpmoore) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/utils/NetNullableElementUtils.java]: These two if statements can be combined into a single one.

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/BaseNullableElement.java (hunk 6 above)
Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/BaseNullableElement.java]: If i add getters, then yes. Other types extend this one, so I need access, but a getter would do that cleaner

Comment (rpmoore) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/elementparsers/BaseNullableElement.java]: Let's do a getter then.

File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/utils/NetNullableElementUtils.java (hunk 7 above)
Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/utils/NetNullableElementUtils.java]: I'm not seeing the simplification

Comment (rpmoore) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/utils/NetNullableElementUtils.java]: You have two nested if statements that can be combined into a single one with an `&&`

Comment (RachelTucker) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/utils/NetNullableElementUtils.java]: But then I'll need to create another if statement for:
if (hasContent(componentType) && isEmpty(encapsulatingXmlTag))

Comment (rpmoore) [inline ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/utils/NetNullableElementUtils.java]: Oh I missed that part.  Ignore this then.

**Comment by RachelTucker** [comment]
Updated
